### PR TITLE
MOTECH-3059 Encounter type will be represented by uuid instead of name

### PR DESCRIPTION
**Pull request by shanmukhm**

### CODE REVIEW

**Comment by kmadej** [comment]
It looks good to me, @shanmukhm did you check OpenMRS ITs after your changes ?

**Comment by shanmukhm** [comment]
@kmadej I ran ITs. I'm getting '400 Bad Request' error (at OpenMRSPatientServiceImpl.createPatient()).
I am getting the same error even if I run ITs on master branch. I think this error is related to configuration and not related to what I changed. Anyway, can you ask someone to run ITs if possible?

**Comment by kmadej** [comment]
Ok no problem, @pkornowski please run ITs with this changes locally

**Comment by pkornowski** [comment]
@shanmukhm @kmadej 
I run ITs locally with this changes and I have 400 as well, but at the first place null pointer appeared in the `MRSTasksIntegrationBundleIT.setUp:157->createEncounterTestData:381->prepareEncounterType:549 NullPointer`.

There is stack trace:

`Tests run: 54, Failures: 0, Errors: 8, Skipped: 0, Time elapsed: 1,087.932 sec <<< FAILURE! - in org.motechproject.openmrs.it.version1_12.MRSIntegrationTests
testOpenMRSPatientDataSourceAndCreatePatientAction(org.motechproject.openmrs.it.MRSTasksIntegrationBundleIT)  Time elapsed: 121.07 sec  <<< ERROR!
java.lang.NullPointerException: null
	at org.motechproject.openmrs.it.MRSTasksIntegrationBundleIT.prepareEncounterType(MRSTasksIntegrationBundleIT.java:549)
	at org.motechproject.openmrs.it.MRSTasksIntegrationBundleIT.createEncounterTestData(MRSTasksIntegrationBundleIT.java:381)
	at org.motechproject.openmrs.it.MRSTasksIntegrationBundleIT.setUp(MRSTasksIntegrationBundleIT.java:157)

testOpenMRSEncounterDataSourceAndCreateEncounterAction(org.motechproject.openmrs.it.MRSTasksIntegrationBundleIT)  Time elapsed: 120.372 sec  <<< ERROR!
org.springframework.web.client.HttpClientErrorException: 400 Bad Request
	at org.springframework.web.client.DefaultResponseErrorHandler.handleError(DefaultResponseErrorHandler.java:88)
	at org.springframework.web.client.RestTemplate.handleResponseError(RestTemplate.java:566)
	at org.springframework.web.client.RestTemplate.doExecute(RestTemplate.java:524)
	at org.springframework.web.client.RestTemplate.execute(RestTemplate.java:495)
	at org.springframework.web.client.RestTemplate.exchange(RestTemplate.java:446)
	at org.motechproject.openmrs.resource.impl.BaseResource.exchange(BaseResource.java:103)
	at org.motechproject.openmrs.resource.impl.BaseResource.postForJson(BaseResource.java:62)
	at org.motechproject.openmrs.resource.impl.PatientResourceImpl.createPatient(PatientResourceImpl.java:44)
	at org.motechproject.openmrs.service.impl.OpenMRSPatientServiceImpl.createPatient(OpenMRSPatientServiceImpl.java:86)
	at org.motechproject.openmrs.it.MRSTasksIntegrationBundleIT.createPatientTestData(MRSTasksIntegrationBundleIT.java:376)
	at org.motechproject.openmrs.it.MRSTasksIntegrationBundleIT.setUp(MRSTasksIntegrationBundleIT.java:156)

testOpenMRSProviderDataSourceAndCreateEncounterAction(org.motechproject.openmrs.it.MRSTasksIntegrationBundleIT)  Time elapsed: 120.53 sec  <<< ERROR!
org.springframework.web.client.HttpClientErrorException: 400 Bad Request
	at org.springframework.web.client.DefaultResponseErrorHandler.handleError(DefaultResponseErrorHandler.java:88)
	at org.springframework.web.client.RestTemplate.handleResponseError(RestTemplate.java:566)
	at org.springframework.web.client.RestTemplate.doExecute(RestTemplate.java:524)
	at org.springframework.web.client.RestTemplate.execute(RestTemplate.java:495)
	at org.springframework.web.client.RestTemplate.exchange(RestTemplate.java:446)
	at org.motechproject.openmrs.resource.impl.BaseResource.exchange(BaseResource.java:103)
	at org.motechproject.openmrs.resource.impl.BaseResource.postForJson(BaseResource.java:62)
	at org.motechproject.openmrs.resource.impl.PatientResourceImpl.createPatient(PatientResourceImpl.java:44)
	at org.motechproject.openmrs.service.impl.OpenMRSPatientServiceImpl.createPatient(OpenMRSPatientServiceImpl.java:86)
	at org.motechproject.openmrs.it.MRSTasksIntegrationBundleIT.createPatientTestData(MRSTasksIntegrationBundleIT.java:376)
	at org.motechproject.openmrs.it.MRSTasksIntegrationBundleIT.setUp(MRSTasksIntegrationBundleIT.java:156)

testCreateEncounterPostActionParameter(org.motechproject.openmrs.it.MRSTasksIntegrationBundleIT)  Time elapsed: 120.316 sec  <<< ERROR!
org.springframework.web.client.HttpClientErrorException: 400 Bad Request
	at org.springframework.web.client.DefaultResponseErrorHandler.handleError(DefaultResponseErrorHandler.java:88)
	at org.springframework.web.client.RestTemplate.handleResponseError(RestTemplate.java:566)
	at org.springframework.web.client.RestTemplate.doExecute(RestTemplate.java:524)
	at org.springframework.web.client.RestTemplate.execute(RestTemplate.java:495)
	at org.springframework.web.client.RestTemplate.exchange(RestTemplate.java:446)
	at org.motechproject.openmrs.resource.impl.BaseResource.exchange(BaseResource.java:103)
	at org.motechproject.openmrs.resource.impl.BaseResource.postForJson(BaseResource.java:62)
	at org.motechproject.openmrs.resource.impl.PatientResourceImpl.createPatient(PatientResourceImpl.java:44)
	at org.motechproject.openmrs.service.impl.OpenMRSPatientServiceImpl.createPatient(OpenMRSPatientServiceImpl.java:86)
	at org.motechproject.openmrs.it.MRSTasksIntegrationBundleIT.createPatientTestData(MRSTasksIntegrationBundleIT.java:376)
	at org.motechproject.openmrs.it.MRSTasksIntegrationBundleIT.setUp(MRSTasksIntegrationBundleIT.java:156)

testCreatePatientPostActionParameter(org.motechproject.openmrs.it.MRSTasksIntegrationBundleIT)  Time elapsed: 120.322 sec  <<< ERROR!
org.springframework.web.client.HttpClientErrorException: 400 Bad Request
	at org.springframework.web.client.DefaultResponseErrorHandler.handleError(DefaultResponseErrorHandler.java:88)
	at org.springframework.web.client.RestTemplate.handleResponseError(RestTemplate.java:566)
	at org.springframework.web.client.RestTemplate.doExecute(RestTemplate.java:524)
	at org.springframework.web.client.RestTemplate.execute(RestTemplate.java:495)
	at org.springframework.web.client.RestTemplate.exchange(RestTemplate.java:446)
	at org.motechproject.openmrs.resource.impl.BaseResource.exchange(BaseResource.java:103)
	at org.motechproject.openmrs.resource.impl.BaseResource.postForJson(BaseResource.java:62)
	at org.motechproject.openmrs.resource.impl.PatientResourceImpl.createPatient(PatientResourceImpl.java:44)
	at org.motechproject.openmrs.service.impl.OpenMRSPatientServiceImpl.createPatient(OpenMRSPatientServiceImpl.java:86)
	at org.motechproject.openmrs.it.MRSTasksIntegrationBundleIT.createPatientTestData(MRSTasksIntegrationBundleIT.java:376)
	at org.motechproject.openmrs.it.MRSTasksIntegrationBundleIT.setUp(MRSTasksIntegrationBundleIT.java:156)

shouldCreateEncounter(org.motechproject.openmrs.it.MRSEncounterServiceIT)  Time elapsed: 0.836 sec  <<< ERROR!
java.lang.NullPointerException: null
	at org.motechproject.openmrs.it.MRSEncounterServiceIT.prepareEncounterType(MRSEncounterServiceIT.java:349)
	at org.motechproject.openmrs.it.MRSEncounterServiceIT.prepareEncounterOne(MRSEncounterServiceIT.java:247)
	at org.motechproject.openmrs.it.MRSEncounterServiceIT.setUp(MRSEncounterServiceIT.java:117)

shouldGetLatestEncounter(org.motechproject.openmrs.it.MRSEncounterServiceIT)  Time elapsed: 0.75 sec  <<< ERROR!
java.lang.NullPointerException: null
	at org.motechproject.openmrs.it.MRSEncounterServiceIT.prepareEncounterType(MRSEncounterServiceIT.java:349)
	at org.motechproject.openmrs.it.MRSEncounterServiceIT.prepareEncounterOne(MRSEncounterServiceIT.java:247)
	at org.motechproject.openmrs.it.MRSEncounterServiceIT.setUp(MRSEncounterServiceIT.java:117)

shouldDeleteEncounter(org.motechproject.openmrs.it.MRSEncounterServiceIT)  Time elapsed: 0.788 sec  <<< ERROR!
java.lang.NullPointerException: null
	at org.motechproject.openmrs.it.MRSEncounterServiceIT.prepareEncounterType(MRSEncounterServiceIT.java:349)
	at org.motechproject.openmrs.it.MRSEncounterServiceIT.prepareEncounterOne(MRSEncounterServiceIT.java:247)
	at org.motechproject.openmrs.it.MRSEncounterServiceIT.setUp(MRSEncounterServiceIT.java:117)


Results :

Tests in error: 
  MRSTasksIntegrationBundleIT.setUp:157->createEncounterTestData:381->prepareEncounterType:549 NullPointer
  MRSTasksIntegrationBundleIT.setUp:156->createPatientTestData:376 » HttpClientError
  MRSTasksIntegrationBundleIT.setUp:156->createPatientTestData:376 » HttpClientError
  MRSTasksIntegrationBundleIT.setUp:156->createPatientTestData:376 » HttpClientError
  MRSTasksIntegrationBundleIT.setUp:156->createPatientTestData:376 » HttpClientError
  MRSEncounterServiceIT.setUp:117->prepareEncounterOne:247->prepareEncounterType:349 NullPointer
  MRSEncounterServiceIT.setUp:117->prepareEncounterOne:247->prepareEncounterType:349 NullPointer
  MRSEncounterServiceIT.setUp:117->prepareEncounterOne:247->prepareEncounterType:349 NullPointer`

**Comment by shanmukhm** [comment]
sorry forgot to update the code yesterday, after this I'm getting '400 Bad request' in ITs (as in master branch)

**Comment by kmadej** [comment]
@shanmukhm i am bit confused with your changes now. I think that the best solution will be using two constructors, one with uuid and one with name. And in this test when the encounterType is creating via REST please use constructor with name.

**Comment by shanmukhm** [comment]
@kmadej we can not add two constructors in EncounterType since both uuid and name are of string.
We can create a new EncounterType only with name and description (according to OpenMRS REST API). UUID is returned after EncounterType is created in openmrs app. So, uuid exists only for already existing encounter type. I think we should not construct with uuid. According to the ticket, when an action of 'create encounter' is selected, user should enter the uuid of existing encounter type.
That's why I only used uuid for task related code. Correct me if I'm wrong.

**Comment by kmadej** [comment]
Hmm, so maybe the best solution will be to create one constructor :
EncounterType(String uuid, String name, String description)

in task action use EncounterType(uuid, null, null) and EncounterTypeSerializer will use only uuid to create encounter

in test while creating encounterType via REST use EncounterType(null, name, description). Json will use only name and description and will ignore null

**Comment by shanmukhm** [comment]
Yeah, that sounds good, will make changes accordingly.

File: openmrs/src/main/java/org/motechproject/openmrs/tasks/impl/OpenMRSActionProxyServiceImpl.java [hunk 1]
```diff
@@ -95,7 +95,7 @@ public Encounter createEncounter(String configName, DateTime encounterDatetime,
         //While creating observations, the encounterDateTime is used as a obsDateTime.
         List<Observation> observationList = MapUtils.isNotEmpty(observations) ? convertObservationMapToList(observations, encounterDatetime) : null;
 
-        EncounterType type = new EncounterType(encounterType);
+        EncounterType type = new EncounterType(encounterType, null);
```
Comment (kmadej) [inline openmrs/src/main/java/org/motechproject/openmrs/tasks/impl/OpenMRSActionProxyServiceImpl.java]: it should be new EncounterType(null, encounterType). Uuid is a second parameter in constructor

**Review by kmadej — CHANGES_REQUESTED** [review]

**Review by kmadej — APPROVED** [review]

**Pull request merged**